### PR TITLE
Greedy (all criterions)

### LINKED ISSUE #10: Greedy algorithm
Implement the greedy algorithm presented

### PR DESCRIPTION
**Pull request by BenoitLoger**
Complete implementation of the greedy algorithm. Some point may be subject to discuss but at least it works. closes #10

### CODE REVIEW

**Review by k-fretaud — APPROVED** [review]
Ready for merge

File: src/greedy.jl [hunk 1]
```diff
@@ -116,7 +116,49 @@ end
 ##===================================================##
 
 """
-    greedy(instance::Instance)
+    filter_on_max_criterion!(candidates::Array{Int64,1}, criterion)
```
Comment (remi-garcia) [inline src/greedy.jl]: drop the `!`

File: src/greedy.jl [hunk 2]
```diff
@@ -116,7 +116,49 @@ end
...
+
+Takes a set of candidates and removes all elements with bad criterion.
+"""
+function filter_on_max_criterion(candidates::Array{Int64,1}, criterion) # TODO: criterion is Int or Float
```
Comment (remi-garcia) [inline src/greedy.jl]: `criterion::Union{Float64, Int}` see [this](https://docs.julialang.org/en/v1/manual/types/#Type-Unions-1) for more info

File: src/greedy.jl [hunk 3]
```diff
@@ -116,7 +116,49 @@ end
...
+    #candidates = tmp_candidates
+    return tmp_candidates
+end
+
+"""
+    filter_on_min_criterion!(candidates::Array{Int64,1}, criterion)
+
+Takes a set of candidates and removes all elements with bad criterion.
+"""
+function filter_on_min_criterion(candidates::Array{Int64,1}, criterion) # TODO: criterion is Int or Float
```
Comment (remi-garcia) [inline src/greedy.jl]: see `filter_on_max_criterion`

File: src/greedy.jl [hunk 4]
```diff
@@ -169,76 +209,81 @@ function greedy(instance::Instance)
...
-                  push!(candidates, tmp_candidates[i])
-              end
+            candidates = filter_on_max_criterion(candidates, tie_break)
+        end
+
+        # Two candidates or more - Second tie break
+        # Cars that require options with higher utilization rates should enter.
+        if length(candidates) > 1
+            # Compute the utilization rate of each options
+            utilization_rate = Array{AbstractFloat,1}(UndefInitializer(),instance.nb_HPRC)
```
Comment (remi-garcia) [inline src/greedy.jl]: Why `AbstractFloat` and not `Float64`?

**Review by remi-garcia — CHANGES_REQUESTED** [review]
A few changes needed and that can be merge.

File: src/greedy.jl [hunk 5]
```diff
@@ -169,76 +209,81 @@ function greedy(instance::Instance)
...
+                        end
+                    end
+                end
+            end
+            candidates = filter_on_min_criterion(candidates, nb_new_violation)
         end
 
-        c = candidates[1]     # We have a valid candidate
-        solution.sequence[pos] = c
+        c = rand([candidates])     # We have a valid candidate
```
Comment (k-fretaud) [inline src/greedy.jl]: `c = rand(candidates)`

`c`should be an integer not an array of integer

**Review by k-fretaud — CHANGES_REQUESTED** [review]

**Review by k-fretaud — DISMISSED** [review]

File: src/greedy.jl [hunk 6]
```diff
@@ -169,76 +207,81 @@ function greedy(instance::Instance)
...
+        # Two candidates or more - LPRC criterion
+        # Choose the car that induces the smallest number of new violations of LPRC
+        # when inserted at the end of the current partial sequence.
         if length(candidates) > 1
-            #TODO
+            # Compute the number of violations caused by each car
+            nb_new_violation = zeros(Int, length(candidates))
+            for ind in 1:length(candidates)
+                c = candidates[ind]
+                for j in (instance.nb_HPRC+1):(instance.nb_LPRC+instance.nb_LPRC)
```
Comment (k-fretaud) [inline src/greedy.jl]: Last option is `(instance.nb_HPRC+instance.nb_LPRC)`

File: src/greedy.jl [hunk 7]
```diff
@@ -169,76 +207,81 @@ function greedy(instance::Instance)
...
+                    if instance.RC_flag[c, j]
+                        for i in ((position - instance.RC_q[j])+1):position
+                            if solution.M1[j, i] >= instance.RC_p[j]
+                                nb_new_violation[c] = nb_new_violation[c] + 1
```
Comment (k-fretaud) [inline src/greedy.jl]: `c` is the car, you want its index `ind` here

File: src/greedy.jl (hunk 4 above)
Comment (BenoitLoger) [reply src/greedy.jl]: Just an habit chomming from other languages, I'll change that

File: src/greedy.jl (hunk 6 above)
Comment (BenoitLoger) [reply src/greedy.jl]: you're right

File: src/greedy.jl (hunk 7 above)
Comment (BenoitLoger) [reply src/greedy.jl]: you're also right

**Review by k-fretaud — APPROVED** [review]

**Review by remi-garcia — APPROVED** [review]

**Pull request merged**